### PR TITLE
Introduce generator/yield-style virtual tables

### PR DESCRIPTION
**Pull request by theopolis**
Here are some worst-case runs from the benchmark:
```
Benchmark                                      Time(ns)    CPU(ns) Iterations
-----------------------------------------------------------------------------
SQL_virtual_table_registry                         1396       1389     509837                                 
SQL_virtual_table_internal                        17014      16957      41442                                 
SQL_virtual_table_internal_yield                  19505      19380      31138                                 
SQL_virtual_table_internal_global                 23149      23075      29675                                 
SQL_virtual_table_internal_unique              16907937   16507892         37                                 
SQL_virtual_table_internal_long                 1239465    1236868        546                                 
SQL_virtual_table_internal_wide/0/1               89877      89634       7112                                 
SQL_virtual_table_internal_wide/0/10             287735     287084       2380                                 
SQL_virtual_table_internal_wide/0/100           2156777    2148746        323                                 
SQL_virtual_table_internal_wide/0/1000         23397962   23321933         30                                 
SQL_virtual_table_internal_wide_yield/0/1         89483      89279       7523                                 
SQL_virtual_table_internal_wide_yield/0/10       268588     267904       2552                                 
SQL_virtual_table_internal_wide_yield/0/100     1979172    1973415        330                                 
SQL_virtual_table_internal_wide_yield/0/1000   19825408   19776879         33
```

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit ae9f905260c00918da9dae44628bc1bbd657f236 (Job results: [1177](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1177/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit ae9f905260c00918da9dae44628bc1bbd657f236 (Job results: [4244](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4244/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit ae9f905260c00918da9dae44628bc1bbd657f236 (Job results: [464](https://jenkins.osquery.io/job/osqueryPullRequestOSX/464/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit ddb58fefddf1867e8fb139f5e53928ffd2f0b464 (Job results: [1178](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1178/)) failed one or more tests (**Windows**).

File: osquery/sql/tests/virtual_table_tests.cpp [hunk 1]
```diff
@@ -421,6 +421,51 @@ TEST_F(VirtualTableTests, test_table_cache) {
   ASSERT_EQ(results[0]["data"], "awesome_data");
 }
 
+class yieldTablePlugin : public TablePlugin {
```
Comment (jedi22) [inline osquery/sql/tests/virtual_table_tests.cpp]: Should this be camel case?

File: osquery/sql/tests/virtual_table_tests.cpp [hunk 2]
```diff
@@ -421,6 +421,51 @@ TEST_F(VirtualTableTests, test_table_cache) {
...
+  EXPECT_EQ(results.size(), 10U);
+  EXPECT_EQ(results[0]["index"], "0");
+
+  results.clear();
+  queryInternal("SELECT * from yield", results, dbc->db());
+  dbc->clearAffectedTables();
+  EXPECT_EQ(results[0]["index"], "10");
+}
+
 class indexIOptimizedTablePlugin : public TablePlugin {
```
Comment (jedi22) [inline osquery/sql/tests/virtual_table_tests.cpp]: Nit: Same

File: osquery/sql/virtual_table.cpp [hunk 3]
```diff
@@ -102,6 +102,14 @@ int xClose(sqlite3_vtab_cursor* cur) {
 
 int xEof(sqlite3_vtab_cursor* cur) {
```
Comment (jedi22) [inline osquery/sql/virtual_table.cpp]: I assume cur cannot or will not be null

File: osquery/sql/virtual_table.cpp [hunk 4]
```diff
@@ -119,6 +127,12 @@ int xDestroy(sqlite3_vtab* p) {
 
 int xNext(sqlite3_vtab_cursor* cur) {
   BaseCursor* pCur = (BaseCursor*)cur;
+  if (pCur->uses_generator) {
```
Comment (jedi22) [inline osquery/sql/virtual_table.cpp]: Same as above; assumed never null

**Comment by theopolis** [comment]
Thanks for the review @jedi22, for the null checks you're correct. We have a promise from the SQLite API where we install those as callbacks. The case management/style is consistent with the other test-tables there, used when class definitions are in mixed order with methods.

**Pull request merged**